### PR TITLE
8341413: Stop including osThread_os.hpp in the middle of the OSThread class

### PR DESCRIPTION
**Pull request by stefank**
Some HotSpot files have an interesting include pattern where the platform dependent code is included straight into the class containing the shared implementation.

This has various interesting effects to the code layout and readability of the code. I propose we stop doing this, where possible, and instead clearly separate the shared code and the platform specific code in separate classes. This then allows us to use the standard include patterns that we use for most of our code.

This PR is a suggestion for how to untangle this for the OSThread class.

Things in the code that changed with this patch that might be good to take an extra look at:
1) I dropped unnecessary includes
2) `pd_initialize/pd_destroy` was converted into constructor/destructor
3) Member initialization lists are used. Note that some variables that used to be uninitialized are now getting initialized. `_caller_sigmask` is one of the interesting once, because it's sizable array. I still don't think that is problematic because all the other code we run to start threads, but if I get push back on this I'll comment it out.
4) (3) switched the order of the `new Monitor` call and the `sigemptyset` call
5) I did some reordering of functions to unify the four platforms
6) Moved `_thread_id` to the platform files
7) I stopped exposing the `thread_id_t` typedef.
8) I created a virtual `thread_id_for_printing` function for those calls that want a unified integer type that can be printed. An alternative to this could be to use a non-virtual call, but that requires us to to define `OSThreadBase::thread_id_for_printing` in the platform files. 

Tested: tier1-3, (excluding AIX, which I can't build/test)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341413](https://bugs.openjdk.org/browse/JDK-8341413): Stop including osThread_os.hpp in the middle of the OSThread class (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [27a9567c](https://git.openjdk.org/jdk/pull/21306/files/27a9567c6c1d3e20833476b373758f78983c1da9)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21306/head:pull/21306` \
`$ git checkout pull/21306`

Update a local copy of the PR: \
`$ git checkout pull/21306` \
`$ git pull https://git.openjdk.org/jdk.git pull/21306/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21306`

View PR using the GUI difftool: \
`$ git pr show -t 21306`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21306.diff">https://git.openjdk.org/jdk/pull/21306.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21306#issuecomment-2388711965)

### CODE REVIEW

File: src/hotspot/share/runtime/osThread.hpp [hunk 1]
```diff
@@ -25,111 +25,7 @@
...
-
-  OSThread();
-  ~OSThread();
-
-  // Printing
-  void print_on(outputStream* st) const;
-  void print() const;
-
-  // Platform dependent stuff
 #include OS_HEADER(osThread)
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/osThread.hpp:29]:
```suggestion

// The actual class declaration is platform specific.
#include OS_HEADER(osThread)
```

**Review by dholmes-ora — APPROVED** [review]
I personally don't have an issue with the current technique to generate a single platform-specific `OSThread` class, but this refactoring is also okay. There is some unfortunate duplication of the boiler-plate code for the class but not too bad.

I'm tempted to suggest pushing the `_startThread_lock` support into `OSThreadBase` under `#ifndef windows`, just to reduce some duplication. (I may also look at using that for Windows too in the near future, which would address it then.)

I could not see where `thread_type()` is actually used so possibly an additional cleanup opportunity there (not necessarily for this PR).

I don't have any concerns with any of the items that you flagged.

Thanks

**Comment by dholmes-ora** [comment]
Aside: forgot to mention, something that seemed odd to me is why we have the OSThread stuff in the os-cpu VMStructs files instead of the os one? I didn't spot anything CPU specific about that.

**Comment by xmas92** [comment]
It is interesting to me that we are using `AllocFailStrategy::RETURN_NULL` when allocating the `OSThread` but then do a `AllocFailStrategy::EXIT_OOM` allocation of a Monitor in the middle of constructing the `OSThread`. 

Even though I am not sure about the usefulness of `AllocFailStrategy::RETURN_NULL`, and whether if it is ever recoverable to fail (smallish) native allocations in Hotspot. Seems like we will crash very soon in any case.

**Comment by dholmes-ora** [comment]
> Even though I am not sure about the usefulness of `AllocFailStrategy::RETURN_NULL`, and whether if it is ever recoverable to fail (smallish) native allocations in Hotspot. Seems like we will crash very soon in any case.

True. The problem with allocations within the constructor is that we have no way to convey to the caller that we had a failure. So if allocation of OSThread fails we act like it is non-fatal because we can just throw  a Java exception from `Thread.start`. But if the constructor fails to construct things properly we have no mechanism to communicate that so we abort on internal allocation failures.

**Comment by xmas92** [comment]
> > Even though I am not sure about the usefulness of `AllocFailStrategy::RETURN_NULL`, and whether if it is ever recoverable to fail (smallish) native allocations in Hotspot. Seems like we will crash very soon in any case.
> 
> The problem with allocations within the constructor is that we have no way to convey to the caller that we had a failure.

Setting a bool field during construction is something we do in a lot of places to signal if the construction was successful.

```c++
  OSThread* osthread = new (std::nothrow) OSThread();
  if (osthread == nullptr || osthread->has_constructor_failed()) {
    delete osthread;
    return false;
  }
```

File: src/hotspot/share/runtime/osThread.hpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/runtime/osThread.hpp:30]: Fixed.

File: src/hotspot/os_cpu/bsd_aarch64/vmStructs_bsd_aarch64.hpp [hunk 2]
```diff
@@ -31,22 +31,9 @@
 // constants required by the Serviceability Agent. This file is
 // referenced by vmStructs.cpp.
 
-#define VM_STRUCTS_OS_CPU(nonstatic_field, static_field, unchecked_nonstatic_field, volatile_nonstatic_field, nonproduct_nonstatic_field, c2_nonstatic_field, unchecked_c1_static_field, unchecked_c2_static_field) \
-                                                                                                                                     \
-  /******************************/                                                                                                   \
-  /* Threads (NOTE: incomplete) */                                                                                                   \
-  /******************************/                                                                                                   \
-  nonstatic_field(OSThread,                      _thread_id,                                      OSThread::thread_id_t)             \
-  nonstatic_field(OSThread,                      _unique_thread_id,                               uint64_t)
```
Comment (coleenp) [inline src/hotspot/os_cpu/bsd_aarch64/vmStructs_bsd_aarch64.hpp:40]: Does the SA actually use these?  I don't see any SA changes in this patch.  If it doesn't, we should remove it until there is someone that might add support for this.  We're not enhancing the SA at this point.

**Review by coleenp — COMMENTED** [review]
Nice cleanup but one question about SA.

File: src/hotspot/os_cpu/bsd_aarch64/vmStructs_bsd_aarch64.hpp (hunk 2 above)
Comment (stefank) [reply src/hotspot/os_cpu/bsd_aarch64/vmStructs_bsd_aarch64.hpp:40]: I think the SA is using these. There are references to them in the various SA files.

**Comment by stefank** [comment]
> I personally don't have an issue with the current technique to generate a single platform-specific `OSThread` class, but this refactoring is also okay.

Thanks for not blocking this suggestion.

> There is some unfortunate duplication of the boiler-plate code for the class but not too bad.

I pushed a change that moved the NONCOPYABLE macro. I did a little experiment with moving even more code to the OSThreadBase, but I'm not sure if this is wanted or not. I've put that change here in-case someone wants to take a look:
https://github.com/openjdk/jdk/compare/pr/21306...stefank:jdk:osThread_follow_up

> 
> I'm tempted to suggest pushing the `_startThread_lock` support into `OSThreadBase` under `#ifndef windows`, just to reduce some duplication. (I may also look at using that for Windows too in the near future, which would address it then.)

I'll leave this for next time someone wants to do a little bit of clean-up in this area.

One interesting thing that we found while looking at this is that the AIX port has the `_startThread_lock`, but it doesn't use it to coordinate the starting of the created thread. It's unclear to me if that's a bug in that port.

> 
> I could not see where `thread_type()` is actually used so possibly an additional cleanup opportunity there (not necessarily for this PR).

Maybe one reason to keep it is as an aid when debugging with a debugger? OTOH, while playing around with the patch linked above I found that the field is not available on Windows, and it's also left uninitialized when we call `create_attached_thread`.

> 
> I don't have any concerns with any of the items that you flagged.
> 
> Thanks

Thanks for reviewing!

File: src/hotspot/os_cpu/bsd_aarch64/vmStructs_bsd_aarch64.hpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/os_cpu/bsd_aarch64/vmStructs_bsd_aarch64.hpp:40]: Ok.  I see now why you don't also have to change SA.

**Review by coleenp — APPROVED** [review]
Looks good. Thanks for removing this inline include.

**Comment by stefank** [comment]
Thanks for reviewing!

**Comment by dholmes-ora** [comment]
> One interesting thing that we found while looking at this is that the AIX port has the _startThread_lock, but it doesn't use it to coordinate the starting of the created thread. It's unclear to me if that's a bug in that port.

Looks like a copy'n'paste bug. AIX starts the new thread in the suspended state the same as Windows. So less sharing of the startThread_lock than I thought at present.

**Comment by dholmes-ora** [comment]
> I found that the field is not available on Windows, and it's also left uninitialized when we call create_attached_thread.

I did some archaeology on this one. The thread_type was introduced way back in May 2000 on Linux as it was needed for the newly started thread to determine whether it had to install (and later remove) an alternate signal stack. When we later stopped using alt-signal-stack (I'm going to guess this was when we switched from LinuxThreads to NPTL) the field was no longer used, but meanwhile we used the ThreadType enum to control initial stack sizes. So Windows never had a thread_type field and BSD/AIX likely just blindly copied it from the Linux code. 

Attaching a thread never involved using alt-signal-stack so for attached threads it was left uninitialized.

So I think we can clean this up - I will file a RFE.

**Pull request closed without merging**